### ISSUE TITLE
Allow selecting distribution metrics to be exposed as summaries instead of histograms

### ISSUE THREAD

**Issue by ivantopo** [kamon-io/kamon-prometheus#13]
Take a look [here](https://gitter.im/kamon-io/Kamon?at=5acf18021130fe3d36cb986d) and a bit bellow.

As discussed by @StephenKing and @kodemaniak there are a few metrics that could or maybe even should be exposed as summaries instead of histograms, maybe even gauges. I personally prefer the idea of summaries cause they can tell a bit more about behavior than a simple gauge but I'm open to discuss it :). These changes would be most certainly applied to JVM and System metrics when exposed to Prometheus.

**Comment by kodemaniak** [comment]
I am perfectly fine with using summaries and I agree that they contain more information than a simple gauge. On the other hand, the simple gauge gives you last measured value. Maybe expose both?

I see the point with aggregating over many nodes, and for latencies and service level metrics in general that's probably required. However, for JVM metrics I think it often makes not much sense to aggregate. Typically, I just need to know which JVM is soon running out of memory (as an example) or behaves badly. Therefore I think aggregating over nodes is not the highest priority for that module.

The biggest problem right now is, that using the system metrics module with Prometheus currently is not intuitive IMHO. You can not use the histograms for monitoring memory usage at all, since the histograms put everything above 1MB into the +Inf bucket. And right now there is now way to change that. Another option could be to let the system metrics module configure the prometheus module with more reasonable bucket sizes and to have some easy config options to customize those defaults (e.g. `system.metrics.jvm.memory=5G`). However, I assume in an ideal world they should not know about each other.

**Comment by jalaziz** [comment]
@ivantopo Spent some time looking into summaries and there is some trickiness that may require changes to kamon-core. The problem is prometheus defines summary quantiles over a sliding time window.

HdrHistogram supports sliding time windows through intervals (see HdrHistogram/HdrHistogram#89). An example of a sliding time window with HdrHistogram can be seen here: https://github.com/HdrHistogram/HdrHistogram/blob/98cacd53338656807d8183ae8bb49e88f66585b0/src/main/java/org/HdrHistogram/HistogramLogProcessor.java#L347-L374.

The issue is that it we do not have access to the underlying HdrHistogram from reporters. We can probably do something similar to `Distribution.merge` but subtracting frequencies in the buckets. However, it could get complicated if we cannot guarantee two distributions have exactly the same bucket values.

All this being said, the current snapshotter behavior may be good enough if we do not use long accumulation periods.

**Comment by ivantopo** [comment]
@jalaziz I totally missed the fact that summaries work on a sliding window!

Regarding implementation, probably we don't need to have access to the HdrHistogram for this because a sliding window will definitely require us to keep all the distributions for the window length and we could `Distribution.merge` them on every tick. For example, if we need to keep the last 10 buckets in the sliding window then every tick will pop the oldest snapshot out of the window and add a new one, so we could:
  - Use `Distribution.merge` to turn these 10 distributions into one and use it to calculate the quantiles for the summary. As a note on this, take a look at how we implemented the [merge method](https://github.com/kamon-io/Kamon/blob/master/kamon-core/src/main/scala/kamon/metric/Distribution.scala#L150-L155), there we are using a `LocalHistogram` to temporarily store data from both distributions and then take a snapshot, we could add all 10 distributions to a single `LocalHistogram` instead and then take a snapshot from it.
  - Besides the 10 distributions, also have an accumulated one for the window and somehow implement a subtract method on it so that on every tick we only do two operations: merge the new distribution and subtract the oldest one. I theory there is no risk of missing buckets because if we encapsulate the window nicely, there is no way to remove a snapshot that wasn't previously added to the accumulated distribution.

WDYT?

**Comment by jalaziz** [comment]
@ivantopo makes sense mostly. I was suggesting exposing the underlying `HdrHistogram` becauses the buffer and subtraction pieces are already in place. Primarily, `HdrHistogram` and therefore `LocalHistogram` has a [`substract` method](https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/AbstractHistogram.java#L782-L804).

My concern with trying to build upon `Distribution.merge` was mainly performance. That is, I'm not sure continuously building a new `LocalHistogram` from 10 distributions just to take a snapshot would be the best option performance-wise. Storing an accumulated distribution alongside the buffer would certainly be better if we can easily subtract from an accumulated distribution. Should mostly be a matter of subtracting counts from buckets, but I haven't looked deeply into the best way to make it work and any edge cases that we may to deal with. Looks like we could reuse `recordValueWithCount`, but the `subtract` method does some extra work.

Are you thinking of keeping the `LocalHistogram` around? That would simplify subtraction, but not sure what the memory implications would be.

My bucketing concern was simply that I'm not sure you can rely on the `bucketsIterator` from two distributions to return the same bucket values, which means we'd need to track which buckets were added and which were removed. However, now that I see that `subtract` uses `recordValueWithCount` my concerns are alleviated.

**Comment by hamishforbes** [comment]
Has there been any progress on this issue?

It seems to me that mapping Kamon's range sampler type to prometheus' histograms does not work very well at all.

Using the trick of `rate(range_sample_sum) / rate(range_sampler_count)` does not produce very accurate metrics.

Is it not possible to map range samplers to 3 prometheus gauges, `current`, `max` and `min`?  
This would seem more intuitive in prometheus and a reasonable compromise?
